docs/mediu/apcm.md

```diff
@@ -187,11 +187,145 @@ int main() {
+            u = dsu.find(edges[i].u);
+            v = dsu.find(edges[i].v);
+            if (u != v) { // sa nu fie in aceeasi componenta
+                if (better(i, minedge[u])) { // cautam cea mai buna muchie pentru fiecare componenta
+                    minedge[u] = i;
+                }
+                if (better(i, minedge[v])) {
+                    minedge[v] = i;
+                }
+            }
+        }
+    }
+}
+
+void processComps() {
+    int i, u, v;
+    for (i = 0; i < n; i++) { // trecem prin fiecare componenta
+        if (minedge[i] != -1 // daca am gasit o muchie pentru componenta in care i e parinte
+                             // daca i nu e parintele unei componente atunci nu o sa fie gasita nicio muchie
+            && viz[minedge[i]] == 0) { // sa nu o fi folosit pentru componenta cu care ne unim deja
+            dsu.join(edges[minedge[i]].u, edges[minedge[i]].v); // unim componentele
+            rez += edges[minedge[i]].cost; // adunam costul
+            viz[minedge[i]] = 1; // am folosit muchia
+        }
+    }
+}
+
+void findMST() {
+    int i, u, v;
+    
+    dsu.init(n);
+    rez = 0;
+    while (dsu.cate_comp > 1) { // cat timp nu e arbore
+        resetComps(); // resetam componentele
+        processEdges(); // trecem prin fiecare muchie
+        processComps(); // unim fiecare componenta cu muchia ei cea mai buna
+    }
+    
+    fout << rez << "\n" << n - 1 << "\n"; // un arbore are n - 1 muchii
+    for (i = 0; i < m; i++) {
+        if (viz[i]) { // daca am folosit muchia o afisam
+            fout << edges[i].u + 1 << " " << edges[i].v + 1 << "\n";
+        }
+    }
+}
+
+int main() {
+    readGraph();
+    findMST();
+    return 0;
+}
+```
+
 ## Care este algoritmul mai bun?
 
-Niciunul dintre algoritmi nu este mai bun mereu decât celălalt. Pe de o parte, Kruskal se dovedește a fi mult mai bun atunci când este vorba de grafuri rare, cu $M \approx N$, deoarece constanta de la sortare este mult mai bună decât cea de la seturi. Totuși, dacă graful este foarte dens, algoritmul lui Prim este superior, iar în cazul unor grafuri complete, de multe ori este mai bine să implementăm varianta sa în $O(n^2)$, similară cu cea prezentată la Dijkstra, pentru a rezolva probleme precum [cablaj](https://www.infoarena.ro/problema/cablaj).
+Niciunul dintre algoritmi nu este mai bun mereu decât celălalt, mai puțin în cazurile particulare în care funcționează doar algoritmul lui Boruvka. Pe de o parte, Kruskal se dovedește a fi mult mai bun atunci când este vorba de grafuri rare, cu $M \approx N$, deoarece constanta de la sortare este mult mai bună decât cea de la seturi. Totuși, dacă graful este foarte dens, algoritmul lui Prim este superior, iar în cazul unor grafuri complete, de multe ori este mai bine să implementăm varianta sa în $O(n^2)$, similară cu cea prezentată la Dijkstra, pentru a rezolva probleme precum [cablaj](https://www.infoarena.ro/problema/cablaj).
```

Comment:
cazuri precum? adica din ce vad, kruskal e m log m si boruvka e m log n dar boruvka pare mai dificil de scris. ar merge aici un exemplu sau ceva in care boruvka sa dea outperform clar la kruskal

Reply:
sau poate o problema in care boruvka bate ceilalti algoritmi